### ISSUE TITLE
Computing loss while onnxrunitme inference 

### ISSUE THREAD

**Issue by OriAlpha** [microsoft/onnxruntime#12526]
This could be a feature, i am trying to compute language model metries. i.e., PERPLEXITY score.
In pytorch it is convenient to get score by 

```
tokenizer = GPT2Tokenizer.from_pretrained('gpt2')
model = GPT2LMHeadModel.from_pretrained('gpt2', return_dict=True)

inputs = tokenizer("here is an example of gpt2 model", return_tensors="pt")
outputs = model(**inputs, labels=inputs["input_ids"])
loss = outputs.loss
```
But in onnx its not clear how can i compute this error?? 

**System information**
- ONNX Runtime version (you are using): 
- Linux: 20.04
- Onnxruntime: 1.12.1
- Python: 3.8

**Comment by tianleiwu** [comment]
I think you will need export a model by adding labels to inputs and loss to the outputs. See corresponding part python:
https://github.com/huggingface/transformers/blob/8cf4a6f0a63ed3aeed68192a9304fed2bd0ce100/src/transformers/models/gpt2/modeling_gpt2.py#L1087-L1094

When the logic is exported to ONNX, you shall see some nodes added to the graph to compute loss based on logits and labels.

You can also export loss calculation to another ONNX model with logits and labels as inputs, and loss as outputs. 

I would probably use the above snippet of python code to calculate the loss directly. That will be much simpler than exporting them to onnx.

**Comment by OriAlpha** [comment]
ok, i modified the class as 

```
class MyGPT2LMHeadModel(GPT2LMHeadModel):
    """Here we wrap a class for Onnx model conversion for GPT2LMHeadModel with past state."""

    def __init__(self, config):
        super().__init__(config)

    def forward(self, labels, input_ids, position_ids, attention_mask, *past):
    #def forward(self, labels, input_ids, position_ids, attention_mask, *past):
        result = super().forward(
            input_ids,
            position_ids=position_ids,
            attention_mask=attention_mask,
            past_key_values=past,
            return_dict=False,
        )
        loss = get_loss(result)

        return MyGPT2Model.post_process(loss, result, self.config.n_layer)
        #return MyGPT2Model.post_process(loss, result, self.config.n_layer)
```
But still its same issue?? Do i have to inport get_loss() function or recreate from transformers??

**Comment by tianleiwu** [comment]
See the code related to result:
https://github.com/huggingface/transformers/blob/8cf4a6f0a63ed3aeed68192a9304fed2bd0ce100/src/transformers/models/gpt2/modeling_gpt2.py#L1096-L1107
The first tuple is the loss. So we can `def get_loss(result): return result[0]`

If you use return_dict=True, the result contains a field called loss.

**Comment by OriAlpha** [comment]
yes, but i need to import into onnxruntime this loss behaviour as i mentioned above?? @tianleiwu

**Comment by tianleiwu** [comment]
@OriAlpha, you will need other changes: For example, create dummy inputs need a real tensor for labels. Need add dynamic axes setting for labels and loss. If you are familiar with torch.onnx.export, it is not hard. You can modify those in gpt2_helper.py (For each place of input_ids, you might need add similar logic for labels input; for each place of logits output, you might need add similar logic for loss output).

**Comment by OriAlpha** [comment]
If possible could @tianleiwu provide a code snippet??, regrading computing loss.
